lms/djangoapps/courseware/date_summary.py

```diff
@@ -240,13 +256,45 @@ def is_enabled(self):
 
     @lazy
     def date(self):
+        deadline = None
+
         try:
-            verified_mode = CourseMode.objects.get(
-                course_id=self.course_id, mode_slug=CourseMode.VERIFIED
-            )
-            return verified_mode.expiration_datetime
+            verified_mode = CourseMode.objects.get(course_id=self.course_id, mode_slug=CourseMode.VERIFIED)
+            deadline = verified_mode.expiration_datetime
         except CourseMode.DoesNotExist:
-            return None
+            pass
+
+        if self.course and self.course_overview.self_paced and self.enrollment:
+            # Check if the given course has opted out of the feature
+            course_config = CourseDynamicUpgradeDeadlineConfiguration.current(self.course.id)
+
+            # NOTE: We check the pk here to determine if we have an instance generated from the current() method or
+            # pulled from the database. If the instance isn't from the database, it is not relevant.
+            if (not course_config.pk) or (course_config.enabled and not course_config.opt_out):
+
+                # Check if dynamic upgrade deadlines are enabled for the site
+                global_config = DynamicUpgradeDeadlineConfiguration.current()
+                if global_config.enabled:
+                    delta = global_config.deadline_days
+
+                    # If the course has an override, use it.
+                    if course_config.enabled:
+                        delta = course_config.deadline_days
+
+                    # The start date is the content availability date--the first date on which the user could
+                    # access the content. This will be the later of either the enrollment date or the
+                    # course's start date.
+                    start = max(self.enrollment.created, self.course_overview.start)
```

Comment:
What should the `content_availability_date` be for users that re-enroll?  I ran into this issue when testing.  I wanted to re-enroll so that my date would get reset.  However, since we used the `created` value, it didn't reset.  Unfo, we may not have a good way of telling the last date the user enrolled.